smack-extensions/src/main/java/org/jivesoftware/smackx/nick/NickManager.java

```diff
@@ -0,0 +1,125 @@
+/**
+ * Copyright 2018 Miguel Hincapie.
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jivesoftware.smackx.nick;
+
+import org.jivesoftware.smack.AsyncButOrdered;
+import org.jivesoftware.smack.Manager;
+import org.jivesoftware.smack.SmackException;
+import org.jivesoftware.smack.StanzaListener;
+import org.jivesoftware.smack.XMPPConnection;
+import org.jivesoftware.smack.filter.StanzaFilter;
+import org.jivesoftware.smack.packet.Message;
+import org.jivesoftware.smack.packet.Stanza;
+import org.jivesoftware.smack.util.StringUtils;
+import org.jivesoftware.smackx.nick.filter.NickFilter;
+import org.jivesoftware.smackx.nick.packet.Nick;
+import org.jxmpp.jid.EntityBareJid;
+
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.WeakHashMap;
+
+/**
+ * @author Miguel Hincapie 2018
+ * @see <a href="http://xmpp.org/extensions/xep-0172.html">XEP-0172: User Nickname</a>
+ */
+public class NickManager extends Manager {
+
+    private static final Map<XMPPConnection, NickManager> INSTANCES = new WeakHashMap<>();
+
+    private static final StanzaFilter INCOMING_MESSAGE_FILTER = NickFilter.INSTANCE;
+
+    private final Set<NickListener> nickListeners = new HashSet<>();
+
+    private final AsyncButOrdered<Message> asyncButOrdered = new AsyncButOrdered<>();
+
+    private NickManager(XMPPConnection connection) {
+        super(connection);
+
+        connection.addSyncStanzaListener(new StanzaListener() {
+            @Override
+            public void processStanza(Stanza packet)
+                    throws
+                    SmackException.NotConnectedException,
+                    InterruptedException,
+                    SmackException.NotLoggedInException {
+                final Message message = (Message) packet;
+
+                asyncButOrdered.performAsyncButOrdered(message, new Runnable() {
```

Comment:
Can you explain me a little bit more about what could be the originating entiy?
Because there is no much doc about that `AsyncButOrdered` I have been using it as I saw in others parts of Smack, so it's not much clear to me how should I use it or when or the options I have instead it.

Reply:
The originating entity is the entity that send the message stanza. So you need to replace `message` with `message.getFrom()`. But make sure to add a null check.

Reply:
and what if it is null?